content/post/2019-10-22-mocking.Rmarkdown

```diff
@@ -0,0 +1,131 @@
+
+```{r mockery}
+# a function that says encoding is a pain
+# when the OS is Windows
+is_encoding_a_pain <- function(){
+  if (Sys.info()[["sysname"]] == "Windows"){
+    return("YES")
+  } else {
+    return("no")
+  }
+}
+
+# The post was rendered on Ubuntu
+Sys.info()[["sysname"]]
+# So, is encoding a pain?
+is_encoding_a_pain()
+
+
+# stub/replace Sys.info() in is_encoding_a_pain()
+# with a mock that answers "Windows"
+mockery::stub(where = is_encoding_a_pain,
+              what = "Sys.info", 
+              how = c(sysname = "Windows"))
+
+# Different output
+is_encoding_a_pain()
+
+# NOT changed
+Sys.info()[["sysname"]]
+```
+
+Let's also look at a real life example, [from `keyring` tests](https://github.com/r-lib/keyring/blob/0cdd366dfd2e8accbf94dd43643531f6f6e1acff/tests/testthat/test-default-backend.R#L56):
+
+```r
+test_that("auto windows", {
+  mockery::stub(default_backend_auto, "Sys.info", c(sysname = "Windows"))
+  expect_equal(default_backend_auto(), backend_wincred)
+})
+```
+
+What happens after the call to `mockery::stub()` is that inside the test, when `default_backend_auto()` is called, it won't use the actual `Sys.info()` but instead a mock that returns `c(sysname = "Windows")` so the test can assess what `default_backend_auto()` returns on Windows... without the test being run on a Windows machine. :sunglasses: 
+
+Instead of directly defining the return value as is the case in this example, one could stub the function with a function, [as seen in one of the tests for the `remotes` package](https://github.com/cran/remotes/blob/f1b3e75c162f555dec0c7ab9dba7dbf9faf69444/tests/testthat/test-install-svn.R#L78).
+
+To find more examples of how to use `mockery` in tests, you can use GitHub search in combination with [R-hub's CRAN source code mirror](https://docs.r-hub.io/#cranatgh): https://github.com/search?l=&q=%22mockery%3A%3Astub%22+user%3Acran&type=Code 
+
+### Web mocking
+
+In the case of a package doing HTTP requests, you might want to test what happens when an error code is received for instance. To do that, you can use either [`httptest`](https://github.com/nealrichardson/httptest) or [`webmockr`](https://github.com/ropensci/webmockr) (compatible with both `httr` and `crul`).
+
+### Temporarily modify the global state
+
+To test what happens when there's, say, an environment variable, one can set it temporarily within a test using the [`withr` package](https://github.com/r-lib/withr). You could argue it's not technically mocking, but it's an useful trick. You can see it in action [in `keyring`'s tests](https://github.com/r-lib/keyring/blob/0cdd366dfd2e8accbf94dd43643531f6f6e1acff/tests/testthat/test-default-backend.R#L18).
```

Comment:
To test what happens when there's, say, an environment variable _has a particular value_ ?

Reply:
Ok, thanks!